pennylane/transforms/batch_partial.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the batch dimension transform.
+"""
+import functools
+import inspect
+
+import pennylane as qml
+
+
+def _convert_to_args(func, args, kwargs):
+    """
+    Given a function, convert the positional and
+    keyword arguments to purely positional arguments.
+    """
+    sig = inspect.signature(func).parameters
+
+    new_args = []
+    for i, param in enumerate(sig):
+        if param in kwargs:
+            # first check if the name is provided in kwargs
+            new_args.append(kwargs[param])
+        elif i < len(sig):
+            # next check if the argnum is provided
+            new_args.append(args[i])
+
+    return tuple(new_args)
+
+
+def batch_partial(qnode, all_operations=False, **partial_kwargs):
+    """
+    Create a wrapper function around the QNode with partially
+    evaluated parameters, which supports an initial batch dimension.
+
+    Args:
+        qnode (pennylane.QNode): QNode to partially evaluate
+        all_operations (bool): If ``True``, a batch dimension will be added to *all* operations
+            in the QNode, rather than just trainable QNode parameters.
+        partial_kwargs (dict): partially-evaluated parameters to pass to the QNode
+
+    Returns:
+        func: Function which accepts the same arguments as the QNode minus the
+        partially evaluated arguments provided, and behaves the same as the QNode
+        called with both the partially evaluated arguments and the extra arguments.
+        However, the first dimension of each argument of the returned function
+        will be treated as a batch dimension. The function output will also contain
+        an initial batch dimension.
+
+    **Example**
+
+    Consider the following circuit:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y[..., 0], wires=0)
+            qml.RY(y[..., 1], wires=1)
+            return qml.expval(qml.PauliZ(wires=0) @ qml.PauliZ(wires=1))
+
+    The ``qml.batch_partial`` decorator allows us to create a partially evaluated
+    function that wraps the QNode. For example,
+
+    >>> y = np.array([0.2, 0.3])
+    >>> batched_partial_circuit = qml.batch_partial(circuit, y=y)
```

Comment:
If I'm reading this correctly, `y` being an array is _independent_ of the batching functionality, right? Could this be confusing for readers? :thinking:

Reply:
Hmmm, you're right, it might be confusing. I'll change the example so that `y` doesn't have to be an array.